plugin/trino-ldap-groups-provider/src/main/java/io/trino/plugin/ldapgroups/LdapGroupsProvider.java

```diff
@@ -0,0 +1,149 @@
+
+    private final LoadingCache<String, Set<String>> groupListCache;
+
+    @Inject
+    LdapGroupsProvider(LdapGroupsProviderConfig config, LdapHighLevelClient ldapHighLevelClient)
+    {
+        requireNonNull(config, "ldap groups provider config is null");
+
+        this.ldapHighLevelClient = ldapHighLevelClient;
+
+        this.groupListCache = EvictableCacheBuilder.newBuilder()
+                .expireAfterWrite(config.getCacheTTL().toMillis(), TimeUnit.MILLISECONDS)
+                .maximumSize(1000)
+                .shareResultsAndFailuresEvenIfDisabled()
+                .build(new CacheLoader<>()
+                {
+                    @Override
+                    public Set<String> load(String user)
+                            throws LdapGroupsLoadingException
+                    {
+                        return getLdapGroups(user);
+                    }
+                });
+    }
+
+    private Set<String> getLdapGroups(String user)
+            throws LdapGroupsLoadingException
+    {
+        ImmutableSet.Builder<String> resBuilder = ImmutableSet.builder();
+        LdapSearchUserResult userSearchResult;
+        try {
+            userSearchResult = ldapHighLevelClient.getUser(user).orElseThrow(() -> new NamingException(String.format("User %s not found in LDAP", user)));
+            log.debug("Fetched user DN and search attributes from LDAP for user %s - %s", user, userSearchResult);
+        }
+        catch (NamingException e) {
+            throw new LdapGroupsLoadingException(String.format("Error loading user %s from LDAP", user), e);
+        }
+
+        resBuilder.addAll(userSearchResult
+                .userAttributes()
+                .entrySet()
+                .stream()
+                .map(e -> e
+                        .getValue()
+                        .stream()
+                        .map(v -> String.format("%s=%s", e.getKey(), v))
+                        .toList())
+                .flatMap(List::stream)
+                .iterator());
+
+        try {
+            List<String> groups = ldapHighLevelClient.getGroups(userSearchResult.userDN());
+            log.debug("Fetched LDAP groups for user %s - %s", user, groups);
+            resBuilder.addAll(groups);
+        }
+        catch (NamingException e) {
+            throw new LdapGroupsLoadingException(String.format("Error loading groups for user %s from LDAP", user), resBuilder.build(), e);
+        }
+
+        return resBuilder.build();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            Set<String> groups;
+            try {
+                groups = groupListCache.get(user);
+            }
+            catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof LdapGroupsLoadingException exception) {
```

Comment:
It would be nice if we could fail explicitly when fetching groups instead of silently returning partial set of groups. WDYT ?

Reply:
IIRC group providers always fail silently since Trino will ignore errors raised by them and continue query execution (or what else it was doing). This is why I though returning _some_ values is better than failing outright.

I am not convinced this is the best approach, so if you think it needs to be changed I'm ok with it

Reply:
Ended up removing this functionality, there was no clean way of using it in conjunction with the `CachingGroupProvider`